components/script/dom/resizeobserver.rs

```diff
@@ -0,0 +1,260 @@
+        proto: Option<HandleObject>,
+        callback: Rc<ResizeObserverCallback>,
+    ) -> DomRoot<ResizeObserver> {
+        let rooted_observer = ResizeObserver::new(window, proto, callback);
+        let document = window.Document();
+        document.add_resize_observer(&rooted_observer);
+        rooted_observer
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#gather-active-observations-h>
+    /// <https://drafts.csswg.org/resize-observer/#has-active-resize-observations>
+    pub fn gather_active_resize_observations_at_depth(
+        &self,
+        depth: &ResizeObservationDepth,
+        has_active: &mut bool,
+    ) {
+        for (observation, target) in self.observation_targets.borrow_mut().iter_mut() {
+            observation.state = Default::default();
+            if observation.is_active(target) {
+                let target_depth = calculate_depth_for_node(target);
+                if target_depth > *depth {
+                    observation.state = ObservationState::Active;
+                    *has_active = true;
+                } else {
+                    observation.state = ObservationState::Skipped;
+                }
+            }
+        }
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#broadcast-active-resize-observations>
+    pub fn broadcast_active_resize_observations(
+        &self,
+        shallowest_target_depth: &mut ResizeObservationDepth,
+    ) {
+        let mut entries: Vec<DomRoot<ResizeObserverEntry>> = Default::default();
+        for (observation, target) in self.observation_targets.borrow_mut().iter_mut() {
+            if !matches!(observation.state, ObservationState::Active) {
+                continue;
+            }
+            // #create-and-populate-a-resizeobserverentry
+            let box_size = calculate_box_size(target, &observation.observed_box);
+            // TODO: writing-mode aware.
+            let width = box_size.width().to_f64_px();
+            let height = box_size.height().to_f64_px();
+            let size_impl = ResizeObserverSizeImpl::new(width, height);
+            let window = window_from_node(&**target);
+            let observer_size = ResizeObserverSize::new(&*window, size_impl);
+            // TODO: padding.
+            let content_rect = DOMRectReadOnly::new(
+                &*window.upcast(),
+                None,
+                box_size.origin.x.to_f64_px(),
+                box_size.origin.y.to_f64_px(),
+                width,
+                height,
+            );
+            let entry = ResizeObserverEntry::new(
+                &*window,
+                target,
+                &*content_rect,
+                // TODO: border box.
+                &[],
+                &[&*observer_size],
+                &[&*observer_size],
+            );
+            entries.push(entry);
+            observation
+                .last_reported_sizes
+                .borrow_mut()
+                .push_back(size_impl);
```

Comment:
I'm using the equivalent that isn't a dom object, so that `ResizeObservation` doesn't have to be a dom object either(which would with our codegen require exposing it to JS).